File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/IconListAdapter.java
package com.remotearthsolutions.expensetracker.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.adapters.viewholder.IconViewHolder;

import java.util.List;

public class IconListAdapter extends RecyclerView.Adapter<IconViewHolder> {

    private List<String> iconlist;
    private IconListAdapter.OnItemClickListener listener;
    private String selectedIcon;
    private GridLayoutManager layoutManager;

    public IconListAdapter(List<String> iconlist, GridLayoutManager layoutManager) {
        this.iconlist = iconlist;
        this.layoutManager = layoutManager;
    }

    @NonNull
    @Override
    public IconViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.custom_icon, parent, false);
        return new IconViewHolder(v, listener, layoutManager);
    }

    @Override
    public void onBindViewHolder(@NonNull IconViewHolder holder, int position) {

        String icon = iconlist.get(position);
        holder.bind(icon, selectedIcon.equals(icon));
    }

    @Override
    public int getItemCount() {

        if (iconlist == null) {
            return 0;
        }
        return iconlist.size();
    }

    public void setOnItemClickListener(IconListAdapter.OnItemClickListener listener) {
        this.listener = listener;
    }

    public void setSelectedIcon(String selectedIcon) {
        this.selectedIcon = selectedIcon;
        int position = iconlist.indexOf(selectedIcon);
        if (position > 0) {
            layoutManager.scrollToPosition(position);
        }
    }

    public interface OnItemClickListener {
        void onItemClick(String icon);
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/IconListAdapter.kt
package com.remotearthsolutions.expensetracker.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.adapters.viewholder.IconViewHolder

class IconListAdapter(
    private val iconlist: List<String>?,
    private val layoutManager: GridLayoutManager
) : RecyclerView.Adapter<IconViewHolder>() {
    private var listener: OnItemClickListener? =
        null
    private var selectedIcon: String? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): IconViewHolder {
        val v =
            LayoutInflater.from(parent.context).inflate(R.layout.custom_icon, parent, false)
        return IconViewHolder(v, listener!!, layoutManager)
    }

    override fun onBindViewHolder(holder: IconViewHolder, position: Int) {
        val icon = iconlist!![position]
        holder.bind(icon, selectedIcon == icon)
    }

    override fun getItemCount(): Int {
        return iconlist?.size ?: 0
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.listener = listener
    }

    fun setSelectedIcon(selectedIcon: String?) {
        this.selectedIcon = selectedIcon
        val position = iconlist!!.indexOf(selectedIcon)
        if (position > 0) {
            layoutManager.scrollToPosition(position)
        }
    }

    interface OnItemClickListener {
        fun onItemClick(icon: String?)
    }

}